### PR TITLE
Better way of handling player while respawning

### PR DESCRIPTION
**Pull request by rourke750**

### CODE REVIEW

**Comment by rourke750** [comment]
@ProgrammerDan @Maxopoly

File: BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/manager/ConnectionManager.java [hunk 1]
```diff
@@ -79,17 +86,48 @@ public boolean teleportPlayerToServer(Player p, String server, PlayerChangeServe
...
-		if (p.isDead()) {
+		if (p.isDead() && !delayedTransit.contains(p.getUniqueId())) {
 			throw new PlayerStillDeadException();
 		}
 		transitManager.addPlayerToExitTransit(p.getUniqueId(), server); // So the player isn't tried to be sent twice.
 		MercuryManager.warnOfArrival(p.getUniqueId(), server); //so target server prepares
 		CustomWorldNBTStorage st = CustomWorldNBTStorage.getWorldNBTStorage();
 		st.save(p, st.getInvIdentifier(p.getUniqueId()), true);
+		
+		p.setGameMode(GameMode.SPECTATOR); // So they cant do anything.
```
Comment (Maxopoly) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/manager/ConnectionManager.java]: Are you sure that the inventory is copied sync, even if saving is done async?

Comment (rourke750) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/manager/ConnectionManager.java]: This is what dan recommended so @ProgrammerDan.
I believe it gets the lock before hand if done async and then saves out and makes the login wait

Comment (ProgrammerDan) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/manager/ConnectionManager.java]: We could add a callback to the async save, and run setgamemode after the async save is complete to address your concern @Maxopoly 

It's a worthy consideration. The lock is grabbed synchronously, so no further saves will run while the lock is held, but if the in-memory representation of inventory was swapped out by GM3 to be empty or whatever, that in-memory representation is what would be asynchronously saved.

**Comment by Maxopoly** [comment]
I dont think hiding other players helps. It makes the amount of abusable situations smaller, but if those exist, this will just make it harder to track down and not prevent it at all.

**Comment by Maxopoly** [comment]
Also for the delayed teleporting, other plugins should just follow the documentation and not do that with dead players.

**Comment by Maxopoly** [comment]
For respawning in the same place see https://github.com/Civcraft/BetterShards/pull/109

**Comment by rourke750** [comment]
Looks good for respawning players in same place.  We need to just make sure we address more situations where things can be abused.  Like players shouldnt be able to do anything while in transit/arriving.

For delayed teleporting I think having bettershards handle it instead of other plugins is better.  Its less overhead.

**Comment by Maxopoly** [comment]
Well so just check in the api teleport method whether the player is dead and if he is schedule it for one tick later

**Comment by Maxopoly** [comment]
Also we should probably add the player to transit in the respawn event, so nothing can happen inbetween the player respawning and the sync runnable to transfer being called

**Comment by rourke750** [comment]
The method has no idea why the player is dead or if it is from the respawnevent

**Comment by rourke750** [comment]
So na to this?

**Pull request closed without merging**